client/modules/sheaf/src/render-handler.ts

```diff
@@ -0,0 +1,123 @@
+import type { Text } from "@codemirror/state"
+import * as FTML from "ftml-wasm-worker"
+import { Memoize } from "typescript-memoize"
+import { toFragment } from "wj-util"
+
+/**
+ * Heavily memoized FTML-render-emit handler.
+ * Designed to invoke the FTML renderer as little as possible.
+ */
+export class RenderHandler {
+  private declare fragmentNode?: DocumentFragment
+
+  /** @param doc - The CodeMirror `state.doc` to render with. */
+  constructor(public doc?: Text) {}
+
+  /**
+   * Raw source of the document.
+   * @decorator `@Memoize()`
+   */
+  @Memoize()
+  get src() {
+    if (!this.doc) return ""
+    return this.doc.toString()
+  }
+
+  /**
+   * Renders the document to CSS and HTML.
+   * @param format - Whether or not to pretty-print/format the HTML.
+   * @decorator `@Memoize()`
+   */
+  @Memoize()
+  async result(format = false) {
+    return await FTML.render(this.src, format)
+  }
+
+  /**
+   * Renders the document to HTML.
+   * @param format - Whether or not to pretty-print/format the HTML.
+   * @decorator `@Memoize()`
+   */
+  @Memoize()
+  async html(format = false) {
+    const { html } = await this.result(format)
+    return html
+  }
+
+  /**
+   * Renders the document's stylesheets.
+   * @decorator `@Memoize`
+   */
+  @Memoize()
+  async styles() {
+    const { styles } = await this.result()
+    return styles
+  }
+
+  /**
+   * Renders the document's combined stylesheet.
+   * @decorator `@Memoize`
+   */
+  @Memoize()
+  async style() {
+    const { styles } = await this.result()
+    return styles.join("\n")
```

Comment:
As [Croque notes here](https://github.com/scpwiki/wikijump/pull/261#discussion_r633123357), stylesheets should be separate because some constructs (e.g. `@import`s) need to be the first thing in the style.